### PR TITLE
DATAUP-314: Job Log Viewer update: add status lines!

### PR DESCRIPTION
**Pull request by ialarmedalien**
# Description of PR purpose/changes

# Jira Ticket / Issue #
e.g. https://kbase-jira.atlassian.net/browse/DATAUP-314
- [x] Added the Jira Ticket to the title of the PR (e.g. `DATAUP-69 Adds a PR template`)

# Testing Instructions
* Details for how to test the PR:
- [x] Tests pass locally and in GitHub Actions
- [x] Changes available by spinning up a local narrative, navigating to an app or bulk import cell, and checking out the logs!

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/lbl.gov/trussresources/home?authuser=0
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] (JavaScript) I have run Prettier and ESLint on changed code manually or with a git precommit hook

### CODE REVIEW

File: kbase-extension/static/kbase/js/util/jobLogViewer.js [hunk 1]
```diff
@@ -687,58 +690,106 @@ define([
          */
         function renderLayout() {
             const uniqueID = html.genId(),
-                events = Events.make(),
-                content = div(
+                events = Events.make();
+            let content = div(
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/util/jobLogViewer.js:694]: git doesn't understand what has happened here. I've changed `content` from a `const` to a `let`; the DOM structure assigned to `content` has not changed.

File: kbase-extension/static/kbase/js/util/jobLogViewer.js [hunk 2]
```diff
@@ -211,6 +214,9 @@ define([
                 {
                     mode: 'completed',
                 },
+                {
+                    mode: 'terminated',
+                },
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/util/jobLogViewer.js:219]: `terminated` was missing from the list of states.

File: kbase-extension/static/kbase/js/util/jobLogViewer.js [hunk 3]
```diff
@@ -687,58 +690,106 @@ define([
...
+                                        class: logContentClass,
+                                    }),
+                                ]
+                            ),
+                        ]
+                    ),
+                ]
+            );
+
+            if (developerMode) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/util/jobLogViewer.js:746]: this is an insert -- in developer mode, some extra divs are added with job info and the widget and FSM state.

File: kbase-extension/static/kbase/js/util/jobLogViewer.js [hunk 4]
```diff
@@ -802,11 +853,119 @@ define([
...
+                div(
+                    {
+                        class: `${cssBaseClass}__job_status_detail`,
+                    },
+                    line
+                )
+            );
+        }
+
+        function renderJobState(jobState) {
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/util/jobLogViewer.js:890]: This function renders the job state (surprisingly!)

If the job state is error, an errorDisplay element is added to show the error details.

File: kbase-extension/static/kbase/js/util/jobLogViewer.js [hunk 5]
```diff
@@ -909,6 +1070,7 @@ define([
                     throw new Error(`Unknown job status ${jobStatus}`);
             }
             if (newState) {
+                renderJobState(lastJobState);
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/util/jobLogViewer.js:1073]: if the job state has changed, update the DOM.

File: kbase-extension/static/kbase/js/util/jobLogViewer.js [hunk 6]
```diff
@@ -986,9 +1152,13 @@ define([
                 key: {
                     type: 'job-log-deleted',
                 },
-                handle: function () {
+                handle: function (message) {
                     stopLogAutoFetch();
                     renderLog();
+                    awaitingLog = false;
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/util/jobLogViewer.js:1158]: if the log has been deleted, we aren't going to have any joy when listening for it.

File: kbase-extension/static/kbase/js/util/jobLogViewer.js [hunk 7]
```diff
@@ -1045,7 +1235,10 @@ define([
         }
 
         function doJobNotFound() {
-            ui.setContent('log-panel', div([p(['No job found; logs cannot be displayed'])]));
+            // clear the log container
+            ui.setContent('log-container', '');
+            renderJobState({ job_state: 'does_not_exist' });
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/util/jobLogViewer.js:1240]: The job status lines will fill in the message about the job not being found.

File: kbase-extension/static/kbase/js/util/jobLogViewer.js [hunk 8]
```diff
@@ -1045,7 +1235,10 @@ define([
...
+            stopJobStatusUpdates();
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/util/jobLogViewer.js:1241]: Obviously if the job is not found, there's no point in getting status updates on it.

File: test/unit/spec/util/jobLogViewerSpec.js [hunk 9]
```diff
@@ -1,24 +1,178 @@
...
+        jobsByStatus[job.status] = job;
+    });
+
+    const endStates = ['completed', 'error', 'terminated'];
+    const queueStates = ['created', 'estimating', 'queued'];
+
+    function getWidgetState(node) {
+        const stateNode = node.querySelector('[data-element="widget-state"]');
+        return JSON.parse(stateNode.textContent);
+    }
```
Comment (ialarmedalien) [inline test/unit/spec/util/jobLogViewerSpec.js:24]: this info shows up in developer mode; the widget is initialised in dev mode for all these tests.

File: test/unit/spec/util/jobLogViewerSpec.js [hunk 10]
```diff
@@ -48,217 +222,493 @@ define(['util/jobLogViewer', 'common/runtime'], (JobLogViewer, Runtime) => {
...
-                        channel: {
+            await this.jobLogViewerInstance.start(arg);
+            const btns = this.node.querySelectorAll('div[data-element="header"] button');
+            btns.forEach((btn) => {
+                expect(btn).toHaveClass('disabled');
+            });
+        });
+
+        describe('initial widget state', () => {
+            describe('should be determining the state with no job object', () => {
```
Comment (ialarmedalien) [inline test/unit/spec/util/jobLogViewerSpec.js:273]: The status line reads 'Determining job state...' if there is no job state object supplied on widget start.

File: test/unit/spec/util/jobLogViewerSpec.js [hunk 11]
```diff
@@ -48,217 +222,493 @@ define(['util/jobLogViewer', 'common/runtime'], (JobLogViewer, Runtime) => {
...
+                            jobState: jobState,
+                        });
+                    });
+                    itHasJobStatus();
+                });
+            });
+
+            JobsData.invalidJobs.forEach((state) => {
+                describe('should be determining the state with a dodgy job state', () => {
+                    const jobId = 'dodgy_job_state_test';
```
Comment (ialarmedalien) [inline test/unit/spec/util/jobLogViewerSpec.js:299]: with an invalid job state object, the status line defaults to being 'Determining job state...'

File: test/unit/spec/util/jobLogViewerSpec.js [hunk 12]
```diff
@@ -48,217 +222,493 @@ define(['util/jobLogViewer', 'common/runtime'], (JobLogViewer, Runtime) => {
...
+                    await this.jobLogViewerInstance.start({
+                        node: this.node,
+                        jobId: 'no state set',
+                    });
+                });
+                itHasJobStatus();
+            });
+
+            JobsData.validJobs.forEach((jobState) => {
+                describe(`should create a string for status ${jobState.status}`, () => {
```
Comment (ialarmedalien) [inline test/unit/spec/util/jobLogViewerSpec.js:284]: Valid job state object supplied -- the appropriate job status line will be produced.

File: test/unit/spec/util/jobLogViewerSpec.js [hunk 13]
```diff
@@ -48,217 +222,493 @@ define(['util/jobLogViewer', 'common/runtime'], (JobLogViewer, Runtime) => {
...
-            const jobId = 'testBtnState';
-            const arg = {
-                node: hostNode,
-                jobId: jobId,
-            };
-            await viewer.start(arg);
-            const btns = hostNode.querySelectorAll('div[data-element="header"] button');
-            btns.forEach((btn) => {
-                expect(btn.classList.contains('disabled')).toBeTruthy();
+        describe('initial widget state, history mode on', () => {
```
Comment (ialarmedalien) [inline test/unit/spec/util/jobLogViewerSpec.js:312]: same as the previous set of tests, but this time with history mode on -- i.e. if the job has gone through the queued and running states, info about queue time and run time will be added to the job status lines.

File: test/unit/spec/util/jobLogViewerSpec.js [hunk 14]
```diff
@@ -48,217 +222,493 @@ define(['util/jobLogViewer', 'common/runtime'], (JobLogViewer, Runtime) => {
...
-                            type: 'job-status',
-                        },
+        /**
+         * Create a MutationObserver that watches for changes to the job status lines
+         *
+         * @param {object} context - jasmine 'this' context
+         * @param {array} jobMessage - appropriately formatted message and channel data to
+         * @returns {Promise} - resolves when there are job status line changes
+         */
+        function createStatusObserver(context, jobMessage) {
```
Comment (ialarmedalien) [inline test/unit/spec/util/jobLogViewerSpec.js:349]: although there are similar functions in the `TestUtil` file -- `waitForElement`, `waitForElementState`, etc. -- this is specialised enough to need its own function

File: test/unit/spec/util/jobLogViewerSpec.js [hunk 15]
```diff
@@ -48,217 +222,493 @@ define(['util/jobLogViewer', 'common/runtime'], (JobLogViewer, Runtime) => {
...
+                            testJobStatus(this, mode);
+                            expect(msg).toEqual({ jobId: jobState.job_id });
+                        });
+
+                        await this.jobLogViewerInstance.start({
+                            node: this.node,
+                            jobId: jobState.job_id,
+                        });
+
+                        return createStatusObserver(this, formatStatusMessage(jobState)).then(
```
Comment (ialarmedalien) [inline test/unit/spec/util/jobLogViewerSpec.js:401]: This promise resolves when there is a change under the job status node -- i.e. the new state is displayed in the UI.

File: test/unit/spec/util/jobLogViewerSpec.js [hunk 16]
```diff
@@ -48,217 +222,493 @@ define(['util/jobLogViewer', 'common/runtime'], (JobLogViewer, Runtime) => {
...
+
+                await this.jobLogViewerInstance.start({
+                    node: this.node,
+                    jobId: jobId,
+                });
+
+                return new Promise((resolve) => {
+                    this.runtimeBus.on('request-latest-job-log', (msg) => {
+                        expect(msg).toEqual({ jobId: jobId, options: {} });
+                        const logUpdate = logs[acc];
```
Comment (ialarmedalien) [inline test/unit/spec/util/jobLogViewerSpec.js:583]: Every time `request-latest-job-log` is called, the next entry in the `logs` array is sent to the widget.

**Review by briehl — CHANGES_REQUESTED** [review]
Just a couple minor tweaks in the jobLogViewer. Tests are very detailed!

File: kbase-extension/static/kbase/js/util/jobLogViewer.js [hunk 17]
```diff
@@ -1065,19 +1260,11 @@ define([
...
             });
-            fsm.bus.on('on-queued', () => {
-                doOnQueued();
-            });
-            fsm.bus.on('exit-queued', () => {
-                doExitQueued();
-            });
             fsm.bus.on('on-job-not-found', () => {
                 doJobNotFound();
+                stopJobStatusUpdates();
```
[the remark this answers is not in the document]
Comment (ialarmedalien) [reply kbase-extension/static/kbase/js/util/jobLogViewer.js:1267]: Fixed.

File: kbase-extension/static/kbase/js/util/jobLogViewer.js [hunk 18]
```diff
@@ -1013,21 +1183,41 @@ define([
...
         }
 
-        function stopEventListeners() {
-            bus.removeListeners(Object.values(listenersByType));
+        /**
+         * Stop the specified listener(s) or all listeners
+         *
+         * If no arguments are supplied, all listeners will be removed
+         *
+         * @param {array{string}} listeners
```
[the remark this answers is not in the document]
Comment (ialarmedalien) [reply kbase-extension/static/kbase/js/util/jobLogViewer.js:1191]: Done.

File: kbase-extension/static/kbase/js/util/jobLogViewer.js [hunk 19]
```diff
@@ -802,11 +853,119 @@ define([
...
+                ].join('\n')
+            );
+
+            if (developerMode) {
+                ui.setContent('dev.job-state', JSON.stringify(jobState, null, 1));
+                renderWidgetState();
+            }
+
+            if (jobState.status === 'running') {
+                runClock = RunClock.make({
```
[the remark this answers is not in the document]
Comment (ialarmedalien) [reply kbase-extension/static/kbase/js/util/jobLogViewer.js:915]: Fixed.

**Review by briehl — APPROVED** [review]
Looks great.

**Pull request merged**